### PR TITLE
feat-validate

### PR DESCRIPTION
**Pull request by lsh-0**
* adds a nginx config file for bot-lax
* adds a uwsgi process for a swagger-backed Flask app

### CODE REVIEW

**Comment by lsh-0** [comment]
work is mostly all there, I can see the config getting tweaked and newrelic being added in the near future

File: salt/lax/config/etc-init-uwsgi-bot-lax-adaptor.conf [hunk 1]
```diff
@@ -0,0 +1,5 @@
+description "uWSGI server"
+start on (local-filesystems and net-device-up IFACE!=lo)
+respawn
+env LANG=en_US.UTF-8
+exec /opt/bot-lax-adaptor/venv/bin/uwsgi --die-on-term --ini /opt/bot-lax-adaptor/uwsgi.ini
```
Comment (giorgiosironi) [inline salt/lax/config/etc-init-uwsgi-bot-lax-adaptor.conf]: we can add New Relic later. I think this doesn't need to run as a particular user, right?

File: salt/lax/config/etc-nginx-sitesenabled-bot-lax.conf [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    return 444;
+}
+{% endif %}
+
+# configuration of the server
+server {
+    # we always listen on port 8001
+    # in non-dev environments port 8001 is only available to internal traffic
+    {% if salt['elife.cfg']('cfn.outputs.DomainName') %}
+    listen      8001 ssl; # see /etc/nginx/nginx.conf for global ssl settings
```
Comment (giorgiosironi) [inline salt/lax/config/etc-nginx-sitesenabled-bot-lax.conf]: probably needs to be opened in `projects/elife.yaml`, at least to internal traffic (plus `bldr update_template` on all lax instances)

File: salt/lax/config/etc-nginx-sitesenabled-bot-lax.conf [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+                {{ salt['elife.cfg']('project.int_project_hostname') }}
+                {{ salt['elife.cfg']('project.int_full_hostname') }};
+    {% else %}
+    listen      8001;
+    server_name localhost;
+    {% endif %}
+
+    charset     utf-8;
+
+    access_log /var/log/nginx/bot-lax.access.log combined_with_time;
```
Comment (giorgiosironi) [inline salt/lax/config/etc-nginx-sitesenabled-bot-lax.conf]: same here for the full naming `bot-lax-adaptor.access.log`, I would keep it consistent

File: salt/lax/config/etc-nginx-sitesenabled-bot-lax.conf [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        uwsgi_pass flask;
+        # drop connection after this many seconds
+        # WARNING: this value *must* be higher than uwsgi's 'harakiri' value (79s) in /srv/lax/uwsgi.ini
+        # at time of writing (2017-06-28) the average scrape time of our largest article was ~40 seconds,
+        # excluding validation and lax round trip
+        uwsgi_read_timeout 80s;
+        # boilerplate, shared with lax uwsgi service
+        include /etc/uwsgi/params;
+
+        # POST/PUT/UPDATE/whatever methods are restricted
```
Comment (giorgiosironi) [inline salt/lax/config/etc-nginx-sitesenabled-bot-lax.conf]: can probably be removed?

File: salt/lax/config/opt-bot-lax-adaptor-uwsgi.ini [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+socket = /tmp/bot-lax-uwsgi.sock
+chmod-socket = 666
+
+logto = /var/log/bot-lax-uwsgi.log
+
+module=wsgi:app
+virtualenv=/opt/bot-lax-adaptor/venv/
+master=True
+chmod-socket = 666
```
Comment (giorgiosironi) [inline salt/lax/config/opt-bot-lax-adaptor-uwsgi.ini]: `chmod-socket` appears multiple times

File: salt/lax/config/etc-nginx-sitesenabled-bot-lax.conf (hunk 2 above)
Comment (lsh-0) [inline salt/lax/config/etc-nginx-sitesenabled-bot-lax.conf]: I simply haven't pushed those changes yet.

File: salt/lax/config/etc-init-uwsgi-bot-lax-adaptor.conf (hunk 1 above)
Comment (lsh-0) [inline salt/lax/config/etc-init-uwsgi-bot-lax-adaptor.conf]: actually, it's uwsgi that does the user switching and not the upstart script. it runs as the www-data user.

File: salt/lax/bot-lax-adaptor.sls [hunk 6]
```diff
@@ -80,3 +80,62 @@ logrotate-for-bot-lax-adaptor-logs:
     file.managed:
         - name: /etc/logrotate.d/bot-lax-adaptor
         - source: salt://lax/config/etc-logrotate.d-bot-lax-adaptor
+
+{% for fname in ['test.log', 'scrape.log', 'validate.log'] %}
```
Comment (lsh-0) [inline salt/lax/bot-lax-adaptor.sls]: yes, `/var/log/bot-lax-adaptor/` perhaps, then I can just target the whole directory and allow them to be created as necessary

File: salt/lax/config/opt-bot-lax-adaptor-uwsgi.ini (hunk 5 above)
Comment (lsh-0) [inline salt/lax/config/opt-bot-lax-adaptor-uwsgi.ini]: fixed

File: salt/lax/config/etc-nginx-sitesenabled-bot-lax.conf (hunk 4 above)
Comment (lsh-0) [inline salt/lax/config/etc-nginx-sitesenabled-bot-lax.conf]: erm, you want the params in there, I'm going to leave the other stuff in there commented out until I've spoken with the consumers of this api and what IPs they're accessing it from. If it's all over the place, I'll put some credentials in place. if it's relatively fixed, we'll limit it by ip range

File: salt/lax/config/etc-nginx-sitesenabled-bot-lax.conf (hunk 3 above)
Comment (lsh-0) [inline salt/lax/config/etc-nginx-sitesenabled-bot-lax.conf]: fixed

**Comment by lsh-0** [comment]
just an aside, a per-app `/var/log/appname` directory feels very convenient. I'm thinking of symlinking the access logs in there are well. 

and as much as I hate to admit it, the unified logging in journalctl in system-d systems *is* convenient. I distrust convenience tradeoffs though.

File: salt/lax/config/etc-nginx-sitesenabled-bot-lax.conf (hunk 4 above)
Comment (giorgiosironi) [inline salt/lax/config/etc-nginx-sitesenabled-bot-lax.conf]: I agree on checking out possible authentication models, but I would avoid the distinction between HTTP methods and just protected all GET/POST/...

**Pull request merged**